llm/src/speed_monitor_w_mfu.py

```diff
@@ -0,0 +1,100 @@
+    described below.
+
+    +----------------------------------+-------------------------------------------------------------+
+    | Key                              | Logged data                                                 |
+    +==================================+=============================================================+
+    |                                  | Rolling average (over ``window_size`` most recent           |
+    | ``throughput/samples_per_sec``   | batches) of the number of samples processed per second      |
+    |                                  |                                                             |
+    +----------------------------------+-------------------------------------------------------------+
+    |                                  | Rolling average (over ``window_size`` most recent           |
+    | ``throughput/mfu``               | batches) of mfu                                             |
+    |                                  |                                                             |
+    +----------------------------------+-------------------------------------------------------------+
+    | ``wall_clock/train``             | Total elapsed training time                                 |
+    +----------------------------------+-------------------------------------------------------------+
+    | ``wall_clock/val``               | Total elapsed validation time                               |
+    +----------------------------------+-------------------------------------------------------------+
+    | ``wall_clock/total``             | Total elapsed time (wall_clock/train + wall_clock/val)      |
+    +----------------------------------+-------------------------------------------------------------+
+
+    Args:
+        window_size (int, optional): Number of batches to use for a rolling average of throughput.
+            Defaults to 100.
+        model_flops (int, optional): Number of FLOPs model uses in the fwd pass per sample. If used, callback will track training MFU.
+            Defaults to None (disabled)
+    """
+
+    def __init__(self, window_size: int = 100, model_flops: int = None):
+        super().__init__(window_size=window_size)
+        self.model_flops = model_flops
+
+    def batch_end(self, state: State, logger: Logger):
+        batch_num_samples = int(state.timestamp.sample) - self.batch_start_num_samples
+        batch_wct = state.timestamp.total_wct.total_seconds() - self.batch_start_wct
+
+        # Add the new element
+        self.batch_wct_buffer.append(batch_wct)
+        self.batch_num_samples_buffer.append(batch_num_samples)
+
+        # Log the throughput
+        if len(self.batch_num_samples_buffer) == self.window_size:
+            throughput = sum(self.batch_num_samples_buffer) / sum(self.batch_wct_buffer)
+            logger.log_metrics({'throughput/samples_per_sec': throughput})
```

Comment:
While we're here... do you want to add these? ;) 

```
throughput/batches_per_sec
throughput/samples_per_sec
throughput/tokens_per_sec
throughput/flops
throughput/device/batches_per_sec
throughput/device/samples_per_sec
throughput/device/tokens_per_sec
throughput/device/flops
throughput/device/mfu
```
(not attached to the naming, just wrote down what first comes to mind. Also would `mfu` be better categorized as a device-level metric?)

Reply:
to get `tokens_per_sec` we end up needing to get pretty specific ie `state.model.model.cfg.max_seq_len`.
This might not generalize well to other settings ie cv or other language models (non-mosaic specific). Is there a better place to get that number?

Reply:
<img width="1413" alt="Screenshot 2022-12-23 at 4 12 26 PM" src="https://user-images.githubusercontent.com/6439018/209414746-e394e45a-3158-4bce-aad5-acf03be93e1f.png">
<img width="1428" alt="Screenshot 2022-12-23 at 4 13 08 PM" src="https://user-images.githubusercontent.com/6439018/209414759-1a34d898-b5ac-4de0-adbd-15994651fa90.png">

Like this?

Reply:
> to get `tokens_per_sec` we end up needing to get pretty specific ie `state.model.model.cfg.max_seq_len`. This might not generalize well to other settings ie cv or other language models (non-mosaic specific). Is there a better place to get that number?

Yeah so there is technically a data spec function called `get_num_tokens_from_batch`, and so we should probably try to call this somehow.. but it doesn't have a smart default so we would need to add a custom data_spec inside `text_data.py`. I can dig into this a bit to see how much work it takes.

Reply:
So I think this line here: https://github.com/mosaicml/composer/blob/2ff5300171b3bd542c0422164ecaa79850869059/composer/callbacks/speed_monitor.py#L104

needs a duplicate line that calls `state.timestamp.token`. And then it's the responsibility of the dataloader + data_spec to properly report the `get_num_tokens_from_batch` to the Trainer. And that would remove the need to use the model's `max_seq_len`.

Reply:
yeah without a dataspec, state.timestamp.token outputs 0:
```
>>> state.timestamp.token
Time(0, TimeUnit.TOKEN)
```
but I guess `state.dataloader.dataset.max_seq_len` is a good place to find seq len (not sure why i didn't just look for it there earlier); we can still check if `max_seq_len` is part of the `state.dataloader.dataset` before adding token info to the logger.
(I guess this wouldn't work for ragged tensors which have variable seq len)
lmk if you still want to updt the dataloader to use a dataspec with the appropriate custom `get_num_tokens_in_batch` fn; its probably the only way to get real counts for variable seq lens if using ragged tensors.